File: Week 4/C++ Codes/ReverseDoublyLinkedList.cpp
#include<bits/stdc++.h>
using namespace std;

struct Node
{
    int data;
    Node * next;
    Node * prev;
    Node (int x)
    {
        data=x;
        next=NULL;
        prev=NULL;
    }
        
};

Node* reverseDLL(Node * head)
{
    struct Node* temp = NULL;
    struct Node* current = head;
    while (current != NULL) {
        temp = current->prev;
        current->prev = current->next; 
        current->next = temp;
        current = current->prev;
    }
        
    if (temp != NULL)
        head = temp->prev;
    return head;
}

File: Week 4/Python Codes/ReverseDoublyLinkedList.py
'''
class Node: 
    def __init__(self, data): 
        self.data = data  
        self.next = None
        self.prev = None
'''

# Pointer swapping
def reverseDLL(head):
    #return head after reversing
    temp = None
    current = head
    while current:
        temp = current.prev
        current.prev = current.next 
        current.next = temp
        current = current.prev
    if temp is not None:
        head = temp.prev
    return head